code/layout-analyzer.js

```diff
@@ -130,22 +199,97 @@ window.addEventListener('DOMContentLoaded', () => {
     table.innerHTML = `<tr><td></td>${cols}<td></td></tr>`;
   };
 
-  // compute the same-finger usages and rolls
-  const computeDigrams = () => {
-    const skuCount = {}; // same-key usage
-    const sfuCount = {}; // same-finger usage
-    const skuDigrams = {};
-    const sfuDigrams = {};
-    const inwardDigrams = {};
-    const outwardDigrams = {};
-    const extendedRolls = {};
-    const scisors = {};
-
-    const fingers = ['l5', 'l4', 'l3', 'l2', 'r2', 'r3', 'r4', 'r5'];
-    fingers.forEach(finger => {
-      sfuCount[finger] = 0;
-      skuCount[finger] = 0;
-    });
+  // Returns a custom iterator, similar to Rust’s std::slice::Windows.
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_generators
+  // https://doc.rust-lang.org/std/primitive.slice.html#method.windows
+  const arrayWindows = (arr, len) => ({
+    *[Symbol.iterator]() {
+      for (let i = 0; i <= arr.length - len; i++) {
+        yield arr.slice(i, i + len);
+      }
+    }
+  });
```

Comment:
OMFG

Reply:
Yeah, in JS fashion the syntax is awful (which is why I added a link to the MDN, glad I did since even you didn’t know that existed), but yeah JS does have some decent stuff in it